### PR TITLE
ActiveStorage::Blobs::ProxyController breaks http1.x chunked responses by setting Content-Length Header

### PR DESCRIPTION
**Pull request by trcarden**
### Summary

Any Downloads larger than 5MB fail in proxy mode for some Http 1.x clients.

ActiveStorage when configured with `resolve_model_to_route = :rails_storage_proxy` will break http 1.x clients due to the additional  "Content-Length" header that is forbidden by [rfc2616 section.4.4](https://datatracker.ietf.org/doc/html/rfc2616#section-4.4).

> ... The Content-Length header field MUST NOT be sent if these two lengths are different (i.e., if a Transfer-Encoding header field is present). ...

**Expected**: A request for a file larger than max chunksize (5MB in the case of ActiveStorage S3) will download successfully on all clients.

**Actual**: Curl, Fastly and other clients hang waiting for the full payload to be received. Eventually the web server persistent connection timeout occurs and download fails. 

### Other Information

The commit that introduced this behavior: https://github.com/rails/rails/pull/41437

Note: Some advanced clients and browsers seem to work intermittently depending their implementation. Chrome for example reports an error on PDFs > 5MB but on retry will complete the download.

There is also a potential issue with `Rack::Response` setting content length on responses that don't have the `Transfer-Encoding = chunked` header. Given this is a protocol dependent decision it doesn't seem like it belongs in Rails, it also may not belong in Rack given `Rack::Chunked` middleware is deprecated. Currently Puma sets the header for you if it detects a missing content-length and the appropriate protocol.

https://github.com/rack/rack/blob/main/lib/rack/response.rb#L334-L342

@tomify I saw your comments on 41437 and I think you might have been experiencing something similar with curl but might not have gotten to the bottom of it before the patch was merged.

### CODE REVIEW

**Comment by arpu** [comment]
@trcarden looks good to me, i get this info from cloudflare on Content-Length headers
but have the same problem you mention and this PR fixed it, hope @tomprats can do a deeper review

**Comment by tomprats** [comment]
It seems right based on what @trcarden said and the rfc, but I no longer have an adequate testing environment to be sure.

> Given this is a protocol dependent decision it doesn't seem like it belongs in Rails

100% agree with this. Wish we could just say `serve file` and then the underlying layers would ensure it could be downloaded however the requester specifies

**Comment by arpu** [comment]
@kaspth can this be merged?

**Comment by feliperaul** [comment]
After a lengthy debug session, I also hit this when using proxy mode.

I initially thought this was related to not having `proxy_http_version 1.1;` in Nginx (which I didnt), or something with Cloudflare. 

However, I then realized I could reliably reproduce it on localhost.

Btw, if disposition is `attachment`, and you open the download pane, you can even see it halts at *exactly* 5mb, and only finishes after a random time has elapsed:

Imagem_colada_14_05_2022_14_14

However, I'm having a somewhat different behavior, and **this PR seems to make things even worse for me**. 

I'm on Rails 7.0.2.3 and Puma 5.6.2, and using Backblaze B2 (which is S3 compatible). 

I'm going to describe 4 scenarios. 

Please use > 10mb PDF files because I found the error even more reliable with > 10mb. Also, enable `Disable cache` in your Chrome devtools and/or use CMD SHIFT R to reload, because the Proxy controller uses `http_cache_forever`.

**Before this PR - Disposition INLINE - CHROME** ❌
This means `response.headers["Content-Length"] = @blob.byte_size.to_s` is set.

On disposition inline, Chrome will display a progress bar halted at the 5mb mark:

image

Then, after a random amount of time, it will fail to load the PDF:

image

**Before this PR - Disposition ATTACHMENT - CHROME** ✅, but only after halting at 5mb
This means `response.headers["Content-Length"] = @blob.byte_size.to_s` is set.

If you add `?disposition=attachment`, Chrome will start the download and halt at the 5mb as well (this is a 12mb file, so progress is a little less than half):

image

After a random amount of time halted, *it will finish the download successfully at once (there is no continuous progress of the download)* and the PDF can be opened normally.

The file size is correct:

image

**After this PR - Disposition INLINE - CHROME** ❌
This means `response.headers["Content-Length"] = @blob.byte_size.to_s` is not set.

There is no progress bar in Chrome indicating that the file is being downloaded (makes sense, since it doesn't know the content-length). However, it will fail to display the PDF every single time with the same "fail to load pdf" message above.

**After this PR - Disposition ATTACHMENT - CHROME** ❌
This means `response.headers["Content-Length"] = @blob.byte_size.to_s` is not set.

Chrome will download the file immediately (after the server responds with 200). However, it will be a corrupt file:

image

FIle size is incorrect at 5.2mb only:

image

Right now, the only thing I can do to make it work reliably with +10mb files is this:

```rb
@blob.open do |tempfile|
  send_file tempfile, disposition: params[:disposition].presence || "inline"
end
```

Which of course is not ideal as it doesn't stream anything, and it has to first download the entire file into a tempfile.

**Firefox**

With Firefox, before this PR, disposition inline worked, because Firefox pdf.js makes multiple GET requests using the `Range` header to request only the pages it needs, so it uses the first branch of that `if` condition (I confirmed this). 

**After this PR, however, it also broke in Firefox the same way I described in Chrome.** 

Lastly, in Firefox, regarding disposition attachment, Firefox itself is to blame, because it currently has a bug (https://bugzilla.mozilla.org/show_bug.cgi?id=1756980) that prevents the download unless you set the `Content-Type` to `application/octet-stream`. 

To fix this while Firefox doesn't fix their bug, we use the `browser` gem and define a singleton method on the blob like this:

```rb
if browser.firefox? && params[:disposition] == "attachment"
  @blob.define_singleton_method(:forcibly_serve_as_binary?) { true }
end
```

It works just fine in files < 5mb; however, files > 5mb don't work with or without this fix, with or without this PR.

@arpu @trcarden @tomprats Could you guys please give it a try with 10+ Mb PDF files? Are you sure this PR is fixing the download (both inline and with disposition attachment) for you?

I'm also using Backblaze B2, I didn't have time to setup a new S3 account to test to see if there was any difference due to this.

**Comment by feliperaul** [comment]
I think I've found the root cause of our problems. 

tl;dr - to fix this, we need to make `ActiveStorage::Streaming` (which is included in the Rails Proxy Controller) a Rails Concern, so it correctly extends `ActionController::Live`'s `module ClassMethod` that it itself includes.

It's due to the fact that `ActiveStorage::Streaming` module, which is included in the `ActiveStorage::Blobs::ProxyController` we're patching in this PR is a regular Ruby module, and not a Rails Concern. So even tough it includes `ActionController::Live`, it won't trigger the `module ClassMethods` of `ActionController::Live` and this makes so that the response is not wrapped in a `Live::Response` instance, which is essential to make all this work. 

This all started when I was doing some tests using the `send_stream` method and discovered that this works just fine:

```rb
class TestController < ApplicationController
  include ActionController::Live  # note this Concern being included
  def download

    send_stream(
      filename: @blob.filename.sanitized,
      disposition: params[:disposition].presence || "inline",
      type: browser.firefox? && params[:disposition] == "attachment" ? "application/octet-stream" : @blob.content_type,
    ) do |stream|
      @blob.download do |chunk|
        stream.write(chunk)
      end
    end

  end

end
```

However, if you swap `ActionController::Live` for `ActiveStorage::Streaming`, you immediately start getting corrupted 5.2mb PDF files that can't be opened.

Digging deeper, `ActiveStorage::Streaming` actually includes `ActionController::Live`, and that's where the `send_stream` method is defined.

However, `ActionController::Live` is a Concern, and it has a `ClassMethods` module with a `make_response!` class method defined:

```rb
module Live
    extend ActiveSupport::Concern

    module ClassMethods
      def make_response!(request)
        debugger
        if request.get_header("HTTP_VERSION") == "HTTP/1.0"
          super
        else
          Live::Response.new.tap do |res|
            res.request = request
          end
        end
      end
    end
    ....
```

So when you include `ActionController::Live` directly, that `make_response` class method gets defined **and I confirmed with a debugger it's being called during the response** ,  thereby wrapping the response in that `Live::Response` instance. This works just fine.

However, when `ActiveStorage::Streaming` is included instead, even tough you are going to have access to `send_stream` method (because it includes `ActionController::Live` so it will be in your controller's ancestors chain), `ActiveStorage::Streaming` **is not a Concern**, being a regular Module instead. So it wont extend the `ActionController::Live` `ClassMethods` module.

Having said all this, note that `ActiveStorage::Blobs::ProxyController` includes `ActiveStorage::Streaming`. So I tried cloning that controller and experimenting with it. Just by including `include ActionController::Live` before the `include ActiveStorage::Streaming` line (or by making `ActiveStorage::Streaming` a concern) the Proxy controller starts working just fine, even with the `send_blob_stream` method.

Lastly, I also confirmed that with this fix the `send_blob_stream` method will work regardless of the `response.headers["Content-Length"]` being set or not, so it appears it wasn't the cause afterall.

I couldn't investigate much further because I had already solved the problem and this took some considerable time to figure out, but note that `Live::Response` wrapper will use `Live::Buffer` that even deletes the Content-length header (check [here](https://github.com/rails/rails/blob/3872bc0e54d32e8bf3a6299b0bfe173d94b072fc/actionpack/lib/action_controller/metal/live.rb#L154)).

**Comment by feliperaul** [comment]
I'm creating a new PR with this fix.

**Comment by feliperaul** [comment]
@trcarden #45102 was merged, can you check if it fixes your issue? If it does, then I think this can be closed.

**Comment by byroot** [comment]
Closing for now, if it doesn't fix your issue let us now we can re-open.

**Comment by trcarden** [comment]
@feliperaul Thank you for taking such a deep look into this. I booted up the old project and gave it a try and it appears to be working without issue after applying your fix.

**Comment by feliperaul** [comment]
@trcarden Glad to hear this Timothy! 🤛🏻
